File: backend/src/models/PublicCheckoutTransaction.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IPublicCheckoutTransaction extends Document {
  plan: 'starter' | 'growth';
  phoneNumber: string;
  amountKes: number;
  darajaRequestId: string;
  status: 'PENDING' | 'SUCCESS' | 'FAILED';
  mpesaReceiptNo?: string;
}

const PublicCheckoutTransactionSchema: Schema = new Schema(
  {
    plan: { type: String, enum: ['starter', 'growth'], required: true },
    phoneNumber: { type: String, required: true, trim: true },
    amountKes: { type: Number, required: true },
    darajaRequestId: { type: String, required: true, unique: true },
    status: { type: String, enum: ['PENDING', 'SUCCESS', 'FAILED'], default: 'PENDING' },
    mpesaReceiptNo: { type: String },
  },
  { timestamps: true }
);

export default mongoose.model<IPublicCheckoutTransaction>(
  'PublicCheckoutTransaction',
  PublicCheckoutTransactionSchema
);

File: backend/app/models/public_checkout_transaction.py
from datetime import datetime, timezone
from typing import Optional, Literal
from pydantic import Field
from app.models.base import BaseDocument


class PublicCheckoutTransaction(BaseDocument):
    plan: Literal["starter", "growth"]
    phone_number: str = Field(alias="phoneNumber")
    amount_kes: float = Field(alias="amountKes")
    daraja_request_id: str = Field(unique=True, alias="darajaRequestId")
    status: Literal["PENDING", "SUCCESS", "FAILED"] = "PENDING"
    mpesa_receipt_no: Optional[str] = Field(default=None, alias="mpesaReceiptNo")
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "publiccheckouttransactions"
